plugins/optimization-detective/class-od-link-collection.php

```diff
@@ -310,6 +320,28 @@ static function ( $matches ) {
 		return 'Link: ' . implode( ', ', $link_headers );
 	}
 
+	/**
+	 * Encode a URL for serving in an HTTP response header.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $url URL to percent encode. Any existing percent encodings will first be decoded.
+	 * @return string Percent-encoded URL.
+	 */
+	private function encode_url_for_response_header( string $url ): string {
+		$decoded_url = urldecode( $url );
+
+		// Encode characters not allowed in a URL per RFC 3986 (anything that is not among the reserved and unreserved characters).
+		$encoded_url = (string) preg_replace_callback(
+			'/[^A-Za-z0-9\-._~:\/?#\[\]@!$&\'()*+,;=]/',
+			static function ( $matches ) {
+				return rawurlencode( $matches[0] );
+			},
+			$decoded_url
+		);
+		return esc_url_raw( $encoded_url );
```

Comment:
Note: This was originally `esc_url_raw( $encoded_url ?? '' )`. However, this seems to have just been a way to work around the fact that `preg_replace_callback()` can technically return `null` when an invalid pattern is provided. However, since we know the pattern is correct based on tests, the use of `(string)` cast seems more appropriate.